c10/util/Logging.h

```diff
@@ -316,6 +316,92 @@ C10_API bool LogAPIUsageFakeReturn(const std::string& context);
 // Initializes the c10 logger.
 C10_API void initLogging();
 
+class C10_API Log {
+ public:
+  Log(const char* component_alias,
+      int64_t py_log_level,
+      const SourceLocation& source_location,
+      std::string msg);
+
+  Log(const char* component_alias,
+      int64_t py_log_level,
+      const SourceLocation& source_location,
+      const char* msg);
+
+  Log(const char* component_alias,
+      int64_t py_log_level,
+      const SourceLocation& source_location,
+      ::c10::detail::CompileTimeEmptyString msg);
+
+  const std::string& component_alias() const;
+  int64_t py_log_level() const;
+  const SourceLocation& source_location() const;
+  const std::string& msg() const;
+
+ private:
+  // The name of the component under which to emit the log message.
+  std::string component_alias_;
+
+  // The Python log level to emit the log as.
+  int64_t py_log_level_;
+
+  // Where the log happened.
+  SourceLocation source_location_;
+
+  // The actual log message.
+  std::string msg_;
+};
+
+// Issue a log with a given message. Dispatched to the current
+// log handler.
+void C10_API log(const Log& log);
+
+class C10_API LogHandler {
+ public:
+  virtual ~LogHandler() = default;
+
+  virtual void process(const Log& log);
+};
+
+namespace LogUtils {
+
+/// Sets the global log handler. This is not thread-safe, so it should
+/// generally be called once during initialization or while holding the GIL
+/// for programs that use python.
+/// User is responsible for keeping the LogHandler alive until
+/// it is not needed.
+C10_API void set_log_handler(LogHandler* handler) noexcept(true);
+
+/// Gets the global log handler.
+C10_API LogHandler* get_log_handler() noexcept(true);
+
+class C10_API LogHandlerGuard {
+  LogHandler* prev_handler_;
+
+ public:
+  LogHandlerGuard(LogHandler* new_handler)
+      : prev_handler_(c10::LogUtils::get_log_handler()) {
+    c10::LogUtils::set_log_handler(new_handler);
+  }
+  ~LogHandlerGuard() {
+    c10::LogUtils::set_log_handler(prev_handler_);
+  }
+};
+
+} // namespace LogUtils
+
 } // namespace c10
 
+// TODO: Combine `TORCH_LOG` with `LOG_AT_FILE_LINE`
+#ifdef DISABLE_LOG
+#define TORCH_LOG(...) ((void)0);
+#else
+#define TORCH_LOG(component_alias, log_level, ...)           \
+  ::c10::log(::c10::Log(                                     \
+      component_alias,                                       \
+      log_level,                                             \
+      {__func__, __FILE__, static_cast<uint32_t>(__LINE__)}, \
+      WARNING_MESSAGE_STRING(__VA_ARGS__)));
```

Comment:
I feel you need to spend more effort thinking about how to avoid doing string formatting when the log isn't actually required. This is one problem with not using a standard logging library like glog. It would be good if you could freely add DEBUG logs and they'd be very cheap (may just one integer test) when disabled.

Reply:
I've thought a little more about this. Right now, the log generated by a `TORCH_LOG(...)` macro call always gets sent from C++ to Python, and then the Python logging system is in charge of filtering it out if its log level is too low. 

The only way I can think of to avoid formatting messages that will not end up being emitted is to move the filtering step down to C++. The C++ side would need to track the log levels of each registered component. That would just require `torch._logging._internal.set_logs` to call some CPython bound function to update the C++ side with the log level state every time it chang
es.

The C++ side filtering step would be more than just a single integer comparison though. It would be a map lookup to find the log level setting of the specified component and then an integer comparison. In the case where the log is filtered out, this would be much faster than it is currently--especially considering that we'd avoid the overhead of emitting the log from C++ to Python.

Does that sound good? I don't know of any better way to do this

Reply:
I also agree that you must move the filtering to C++. It might be worth thinking through some potential ways to make the level test as cheap as possible; speed of light is probably a memory dereference and an integer test, but that seems difficult to achieve without setting some sort of per-file global variable and then registering it in a way that `set_logs` knows how to toggle it. Maybe that is an acceptable UX compromise. Another choice that no one could object to is to cost as much as glog does.